### DISCUSSION TITLE
blog/2021/12/20/fully-bayesian-ate-iptw/index

### DISCUSSION THREAD

**Discussion by giscus** [andrewheiss/ath-quarto#64] [Blog comments]
# blog/2021/12/20/fully-bayesian-ate-iptw/index

Use a posterior distribution of inverse probability weights in a Bayesian outcome model to conduct (nearly) fully Bayesian causal inference with R, brms, and Stan

https://www.andrewheiss.com/blog/2021/12/20/fully-bayesian-ate-iptw/index.html

**Comment by StaffanBetner** [comment]
This approach can also be used for ("Bayesian"/fractional-random-weight) bootstrapping by sampling weights from X, where X/n is Dirichlet distributed, typically with the vector of alphas consisting only of ones.

**Comment by StaffanBetner** [comment]
To make it work with `cmdstanr`, add this line in the beginning of `iterfuns.hpp`:
> #include \<ostream\>

Then you just compile the model like this:
> cmdstan_model("modified_stan_code.stan", user_header = here('iterfuns.hpp')) -> modified_model

**Comment by joscani** [comment]
wow. 
Great post. 
I didn't know that ipw works in bayesian framework. Thanks

**Comment by joscani** [comment]
Other alternative is use "full luxury bayes" .In Bayesian we can fit all dag jointly. 

Here is the code for thist dag. 


```
# Full bayesian models without IPTW

bf_treatment <-  bf(net ~ income + temperature + health ,
      family = bernoulli())

bf_health <- bf(health  ~  income, family = gaussian())

bf_outcome <-  bf(malaria_risk  ~ net + income + temperature + health , family = gaussian())

fit2 <- brm(bf_treatment + bf_outcome + bf_health, 
            chains = 4, cores = 4, iter = 2000, data = nets,
            seed = 1234, backend = "cmdstanr")


summary(fit2)
```
And `marlariarisk_netTRUE` is the ATE , near to the true effect without using IPW.  This is luxury because is expensive in computation, But in this mindset we fit all  dag relationship 


``` 
Population-Level Effects: 
                        Estimate Est.Error l-95% CI u-95% CI Rhat
net_Intercept              -1.33      0.38    -2.06    -0.58 1.00
malariarisk_Intercept      76.16      0.96    74.32    78.01 1.00
health_Intercept          -15.87      1.35   -18.51   -13.26 1.00
net_income                  0.00      0.00     0.00     0.00 1.00
net_temperature            -0.06      0.01    -0.08    -0.04 1.00
net_health                  0.01      0.00    -0.00     0.02 1.00
malariarisk_netTRUE       -10.44      0.27   -10.96    -9.91 1.00
malariarisk_income         -0.08      0.00    -0.08    -0.07 1.00
malariarisk_temperature     1.01      0.03     0.94     1.07 1.00
malariarisk_health          0.15      0.01     0.13     0.17 1.00
health_income               0.07      0.00     0.07     0.08 1.00
                        Bulk_ESS Tail_ESS
net_Intercept               4237     3548
malariarisk_Intercept       4005     2781
health_Intercept            5695     2807
net_income                  5245     3516
net_temperature             4064     2901
net_health                  6218     3085
malariarisk_netTRUE         4150     3206
malariarisk_income          3717     2830
malariarisk_temperature     3945     2608
malariarisk_health          3511     2885
health_income               5726     2625
```

The original idea was from 
Richard McElreath . Here you are a post about this. https://elevanth.org/blog/2021/06/29/regression-fire-and-dangerous-things-3-3/

**Reply by joscani** [reply]
My approximation was modelling the entire dag. In this case is similar to use backdoor criterio from pearl, and ipw is a different thing.  Anyway, modelling complete dag is useful If we need to control by unobserved confound or ¡condition by a collider! . Yes in full luxury bayesian we can control by a collider.

**Comment by martinmodrak** [comment]
Awesome, I was just thinking about the best way to do IPTW in a Bayesian way and this + the previous post answers most of my questions.

One thing I wonder is that it seems to me that you should be able to treat the selection and outcome model jointly (though likely not easily with `brms`). I've been working with selection models independently, but Mike Betancourt has recently published an intro that's quite good (https://betanalpha.github.io/assets/chapters_html/modeling_selection.html) and the `1/p(z = 1)` term in section 1.1 seems to be exactly analogous to the IPTW (for context, the example in Mike's post is IMHO best understood as describing the hypothetical population of subjects with exactly identical values of all confounders - you thus need separate `p(z = 1)` for each subject). This leaves you with the difficult task of evaluating the integral to compute `p(z = 1)` analytically OR biting the bullet and treating `p(z = 1)`  (which seems to correspond to `\nu` in your blog post) as a parameter, i.e. add a parameter for each observation.

Are you aware of people trying to make this work?

**Comment by pbilodeau94** [comment]
Is there a way to adapt this for multiple treatments rather than binary?
